### ISSUE TITLE
Give the element that gets focused when navigating with j and k a sensible aria-label for screen readers.

### ISSUE THREAD

**Issue by MarcoZehe** [tootsuite/mastodon#8192]
Right now, if using a screen reader in non-virtual mode so one can use the J and K keys to navigate up and down, and 1 through 3 to quickly switch columns, the screen reader NVDA on Windows with Firefox will only speak "document x of y, section" when landing on a new item. On the Mac, because VoiceOver treats the container as something to grab all child text from, it will read something in the order it appears in in the DOM, but it's actually not the most pleasant user experience.

For efficiency when reading, I suggest to add an aria-label attribute to the div tha tgets focus with j and k, and add information in this order, constructed from the toot's data:

- Author screen name, followed by a comma and a space,
- Toot text with all punctuation left intact, followed by another comma and a space. Don't worry if punctuation is contained at the end, screen readers will handle this.
- The time stamp, like "1 minute ago", followed by another comma and a space,
- The username including instance.
- If this is a boost, another comma and space, and then text like "boosted by ", followed by whoever boosted the toot

Granted, the order could be argued about, but over the years, I found this to be the order of information most blind users I talked to) found most useful, based on similar clients for Twitter and other social media platforms.

For example, [this toot of mine](https://toot.cafe/@marcozehe/100542795612909474) would be read:

Marco Zehe, I got back from vacation today. Caught up on Bugmail and other stuff. I brought my build system up to speed. I was hyper-focused all morning. And now I feel a bit like a balloon that the air was let out of., 23h ago, @marcozehe@toot.cafe

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by MarcoZehe** [comment]
Forgot to add: For notifications, this could be easier, since the text is already there, and almost in the right order. The section should get an aria-labelledby here with IDrefs pointing to the "xy favorited your status" and the bits that make up the status. They'll grab the inner text automatically, but you could definitely use the Firefox accessibility inspector to check whether the text looks reasonable. It'll be the name property that's relevant here.

**Comment by pitermach** [comment]
As a user, I don't really find it too hacky. This practice seems to have been adopted by a lot of sites, including Twitter for instance. Judging by what @MarcoZehe   said, browsers that build a usable accessibility name by looking at the DOM aren't doing the correct standard behaviour. To add to his findings, using Microsoft Edge with Narrator just says "group" without reading the toot. When using Chrome with NVDA on Windows, toots get read out, but differently to Safari. For some reason the time is not spoken at all, but for every toot it always announces the labels of every button. So every toot gets wrapped up with the screen reader saying "reply boost fav more." and you can't just not expose them in the DOM at all because then if someone is navigating the page with the screen reader commands, rather than J and K, they won't be able to see the controls to interact with a toot.

**Comment by mike-burns** [comment]
> This practice seems to have been adopted by a lot of sites, including Twitter for instance.

It looks like Twitter uses an `aria-labelled-by`, not `aria-label`. That seems like a much less hacky solution. Is that possible to use here, too?

**Comment by MarcoZehe** [comment]
Yes, you could use aria-labelledby and reference the items in the wanted order that contain author, text, date etc., in the most sensible order. The downside is that you don't get control over the separators. By default, text of those items only gets separated by a single space. The advantage of aria-label and a made-up string would be that you could insert pauses by inserting commas, after the author name, and some other places, to give the items just a little separation, sort of like taking a silent breath.

**Comment by MarcoZehe** [comment]
> I'm assuming it doesn't make sense to have the aria-label change based on the "show more"/"show less" button

That is exactly what should happen. Only add the CW text if it is hidden and the full text instead when it is showing.

**Comment by pitermach** [comment]
A couple of things I noticed:

- The most minor is that the timestamps that are displayed next to toots are relative, while the ones added to the label are absolute. Personally, I'd prefer to hear the relative timestamps, which would make more sense as it matches with what's being displayed.
- For toots that have a content warning, I think there should be an indication that the toot is collapsed. Sometimes I don't find it obvious that what I just read is actually a content warning. This indication could either just be "Content warning:" added before the warning, or perhaps saying "post is hidden" added after the warning.
- There is no indication at all of attached media, including images that have alt tags. In general, I think saying "image" or "video" for toots that have media should happen, but more importantly if the attached image has an alt tag it should be included in the label.
- For the notification column, anything that's not a toot still has no label which results in Firefox reading nothing at all for those items. This includes follows, boosts and favorites, at least that's all the ones I saw.

**Comment by MarcoZehe** [comment]
@Gargron Regarding the items that @pitermach mentioned above, should we add issues for the image description, CW and time stamp items? I think these are valid points that should be added as a refinement to the new labels.
